sql/expression/function/aggregation/window/lag.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package window
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"gopkg.in/src-d/go-errors.v1"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+)
+
+var ErrInvalidLagOffset = errors.NewKind("'LAG' offset must be a non-negative integer; found: %v")
+var ErrInvalidLagDefault = errors.NewKind("'LAG' default must be a literal; found: %v")
+
+type Lag struct {
+	window *sql.Window
+	expression.NaryExpression
+	//def sql.Expression
+	offset int
+	pos    int
+}
+
+var _ sql.FunctionExpression = (*Lag)(nil)
+var _ sql.WindowAggregation = (*Lag)(nil)
+
+// getLagOffset extracts a non-negative integer from an expression.Literal, or errors
+func getLagOffset(e sql.Expression) (int, error) {
+	lit, ok := e.(*expression.Literal)
+	if !ok {
+		return 0, ErrInvalidLagOffset.New(e)
+	}
+	val := lit.Value()
+	var offset int
+	switch e := val.(type) {
+	case int:
+		offset = e
+	case int8:
+		offset = int(e)
+	case int16:
+		offset = int(e)
+	case int32:
+		offset = int(e)
+	case int64:
+		offset = int(e)
+	default:
+		return 0, ErrInvalidLagOffset.New(e)
+	}
+
+	if offset < 0 {
+		return 0, ErrInvalidLagOffset.New(e)
+	}
+
+	return offset, nil
+}
+
+// NewLag accepts variadic arguments to create a new Lag node:
+// If 1 expression, use default values for [default] and [offset]
+// If 2 expressions, use default value for [default]
+// 3 input expression match to [child], [offset], and [default] arguments
+// The offset is constrained to a non-negative integer expression.Literal.
```

Comment:
It can also apparently be a user var, add a TODO?

Reply:
added TODO